packages/react-dom-bindings/src/server/fizz-instruction-set/ReactDOMFizzInstructionSetShared.js

```diff
@@ -125,3 +125,84 @@ export function completeSegment(containerID, placeholderID) {
   }
   placeholderNode.parentNode.removeChild(placeholderNode);
 }
+
+// This is the exact URL string we expect that Fizz renders if we provide a function action.
+// We use this for hydration warnings. It needs to be in sync with Fizz. Maybe makes sense
+// as a shared module for that reason.
+const EXPECTED_FORM_ACTION_URL =
+  // eslint-disable-next-line no-script-url
+  "javascript:throw new Error('A React form was unexpectedly submitted.')";
+
+export function listenToFormSubmissionsForReplaying() {
+  // A global replay queue ensures actions are replayed in order.
+  // This event listener should be above the React one. That way when
+  // we preventDefault in React's handling we also prevent this event
+  // from queing it. Since React listens to the root and the top most
+  // container you can use is the document, the window is fine.
+  // eslint-disable-next-line no-restricted-globals
+  addEventListener('submit', event => {
+    if (event.defaultPrevented) {
+      // We let earlier events to prevent the action from submitting.
+      return;
+    }
+    const form = event.target;
+    const submitter = event['submitter'];
```

Comment:
do we need better closure externs? (do we even get anything out of advanced mode?)

Reply:
Yea, I added one for addEventListener already. Not sure we need advanced mode tbh.

Reply:
I think the only reason I made it use advanced was to avoid having to run Rollup on it to minify the symbols (though there's probably a better way to do that besides advanced mode) but it looks like we're probably going to have to run Rollup on this file anyway

Reply:
Ideally it would use the same build pipeline as every other file but there's a sequencing issue because our normal build pipeline doesn't have an easy way to do codegen. We'd have to run the build script twice with different arguments, maybe.